execution/src/vm.rs

```diff
@@ -7,15 +7,74 @@ use wasmer::{imports, Function, ImportObject, Instance, Module, Store, WasmerEnv
 use crate::sce_ledger::{SCELedger, SCELedgerChanges, SCELedgerStep};
 use crate::ExecutionConfig;
 
-/// Example API available to wasm code, aka "syscall".
+/// Example ABI available to wasm code, aka "syscall".
 fn foo(shared_env: &SharedExecutionContext, n: i32) -> i32 {
     n
 }
 
-/// API allowing a contract to call another.
-fn call_address(shared_env: &SharedExecutionContext, addr: Address) {
-    //if let Some(ledger_entry) = Arc::clone(&shared_env.0).lock().active_ledger.get(&addr) {}
-    //TODO
+/// ABI allowing a contract to call another.
+fn call(shared_env: &SharedExecutionContext, addr: Address, func_name: String, max_gas: u64) {
+    //TODO add arbitrary input parameters and return value
+
+    //TODO metering / mem limit
+
+    // prepare execution
+    let old_max_gas;
+    let old_coins;
+    let target_module;
+    let ledger_push;
+    {
+        let mut exec_context_guard = shared_env.0.lock().unwrap();
+
+        // TODO make sure max_gas >= context.remaining_gas
+
+        // get target module
+        if let Some(module) = (*exec_context_guard).ledger_step.get_module(&addr) {
+            target_module = module;
+        } else {
+            // no module to call
+            // TODO error
+            return;
+        }
+
+        // save old context values
+        ledger_push = (*exec_context_guard).ledger_step.caused_changes.clone();
+        old_max_gas = (*exec_context_guard).max_gas; // save old max gas
+        old_coins = (*exec_context_guard).coins;
+
+        // update context
+        (*exec_context_guard).max_gas = max_gas;
+        (*exec_context_guard).coins = 0; // TODO maybe allow sending coins in the call
```

Comment:
From what I understood, we really want to send coins in order to pay the smart contract.

Reply:
Here we're only calling it from the context of another SC. But yeah, we could add a "coins" system to make coins available for some executions. I haven't done it yet, hence the "TODO"